File: drone_dob/include/utils/LowPassFilter.cpp
#include "LowPassFilter.hpp"

LowPassFilter::LowPassFilter()
: cutoff_frequency_(50.0) 
{

}

LowPassFilter::LowPassFilter(double cutoff_freq)
: cutoff_frequency_(cutoff_freq)
{

}

void LowPassFilter::setCutoffFrequency(double cutoff_freq) {
    cutoff_frequency_ = cutoff_freq;
}

double LowPassFilter::update(double input, double dt) {
    double alpha = 1 - exp(-dt * 2 * M_PI * cutoff_frequency_);
    // std::cout << "Alpha: " << alpha << std::endl;
    // std::cout << "dt: " << dt << std::endl;
    output_ = output_*(1 - alpha) + input * alpha;
    return output_;
}

double LowPassFilter::getOutput() const {
    return output_;
}

File: drone_control/drone_control/utils/low_pass_filter.py
import numpy as np

class LowPassFilter(object):
    """First-order discrete LPF: y[k] = β y[k-1] + (1-β) x[k],  β = exp(-2π fc dt).

    The output vector is initialized lazily on the first filter() call to match
    the shape of the input, so a single class works for scalar, ω (3-d), full
    odom (6/13-d), etc.
    """

    def __init__(self, cutoff_freq=10.0):
        self.cutoff_freq = cutoff_freq
        self.output = None    # initialized on first filter() call

    def reset(self, value):
        self.output = np.asarray(value, dtype=float).copy()

    def filter(self, input, dt):
        x = np.asarray(input, dtype=float)
        if self.output is None or self.output.shape != x.shape:
            self.output = x.copy()
            return self.output
        if self.cutoff_freq <= 0.0:
            self.output = x.copy()    # bypass: cutoff disabled
            return self.output
        beta = np.exp(-2.0 * np.pi * self.cutoff_freq * dt)
        self.output = beta * self.output + (1.0 - beta) * x
        return self.output